worker/uniter/operation/runaction.go

```diff
@@ -78,22 +88,46 @@ func (ra *runAction) Execute(state State) (*State, error) {
 		// be handled inside the Runner, and returned as nil.
 		return nil, errors.Annotatef(err, "running action %q", ra.name)
 	}
-	return stateChange{
-		Kind:     RunAction,
-		Step:     Done,
-		ActionId: &ra.actionId,
-		Hook:     state.Hook,
-	}.apply(state), nil
+
+	var newState *State
+	if !ra.nonBlocking {
+		newState = stateChange{
+			Kind:     RunAction,
+			Step:     Done,
+			ActionId: &ra.actionId,
+			Hook:     state.Hook,
+		}.apply(state)
+	}
+
+	return newState, nil
 }
 
 // Commit preserves the recorded hook, and returns a neutral state.
 // Commit is part of the Operation interface.
 func (ra *runAction) Commit(state State) (*State, error) {
```

Comment:
Updating the state on commit is likely going to need to be made concurrent or action ids will mysteriously begin to reappear when multiple actions complete at around the same time.